File: data_structures/queue/queue.cpp
using namespace std;
#include <iostream>

/**
 * Defines a node in a Queue.
 *
 * Only supports integer values.
 */
class Node {
  public:
    int value;
    Node* next;

    Node(int value, Node* next=NULL) {
      this->value = value;
      this->next = next;
    }
};

/**
 * Defines a linked list based queue data structure.
 *
 * Time complexity: O(1)
 */
class Queue {
  public:
    Queue() {
      this->head = NULL;
      this->tail = NULL;
    }

    int peek() {
      return this->head ? this->head->value : -1;
    }

    int pop() {
      Node* temp = this->head;
      this->head = this->head->next;

      int buffer = temp->value;
      delete temp;

      return buffer;
    }

    void push(int value) {
      Node* node = new Node(value);

      // if head has not been set yet, set it.
      if (!this->head) {
        this->head = node;
      }

      // if tail has been set, set it's next to the new node
      if (this->tail) {
        this->tail->next = node;
      }

      this->tail = node;
    }

  private:
    Node* head;
    Node* tail;
};

int main() {
  Queue queue;
  queue.push(1);
  queue.push(2);
  queue.push(3);
  queue.push(4);

  cout << queue.pop() << endl;
  cout << queue.pop() << endl;
  cout << queue.pop() << endl;
  cout << queue.peek() << endl;
  cout << queue.pop() << endl;
  cout << queue.peek() << endl;
  return 0;
};

File: data_structures/queue/queue.go
package main

import "fmt"

type Node struct {
  value int
  next *Node
}

func NewNode(value int, next *Node) *Node {
  node := new(Node)
  node.value = value
  node.next = next

  return node
}

type Queue struct {
  head *Node
  tail *Node
}

func NewQueue(head *Node, tail *Node) *Queue {
  queue := new(Queue)
  queue.head = head
  queue.tail = tail

  return queue
}

func (q *Queue) peek() int {
  if (q.head != nil) {
    return q.head.value
  } else {
    return -1
  }
}

func (q *Queue) pop() int {
  temp := q.head
  q.head = q.head.next

  value := temp.value

  return value
}

func (q *Queue) push(value int) {
  node := NewNode(value, nil)

  // if head has not been set
  if (q.head == nil) {
    q.head = node
  }

  // if tail has been set, set it's next to the new node
  if (q.tail != nil) {
    q.tail.next = node
  }

  q.tail = node
}

func (q *Queue) printQueue() {
  temp := q.head

  for temp != nil {
    fmt.Println(temp.value)
    temp = temp.next
  }
}

func main() {
  queue := NewQueue(nil, nil)

  queue.push(4)
  queue.push(2)
  queue.push(42)

  queue.printQueue()

  fmt.Println(queue.pop())
  fmt.Println(queue.pop())
  fmt.Println(queue.peek())
  fmt.Println(queue.pop())
  fmt.Println(queue.peek())
}
